File: 15_pde_fdm/problems/01_wave/wave.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <omp.h>

/* 2D 波動方程式 u_tt = c^2 (u_xx + u_yy) を陽解法で時間発展させる。
   L×L の膜。四辺を固定 (u=0) し, 中央に山(ガウス)を置いて波が広がり反射する様子を計算する。
   更新式 (5点ラプラシアン, coef = (c*dt/dx)^2, 安定条件 coef <= 0.5):
     u^{n+1} = 2 u^n - u^{n-1} + coef * (上+下+左+右 - 4*中央)
   時間方向は前後のステップに依存するので逐次, 空間の二重ループを並列化する。 */
int main(int argc, char ** argv) {
  int    L     = (argc > 1 ? atoi(argv[1]) : 257);     /* 一辺の格子点数 */
  int    steps = (argc > 2 ? atoi(argv[2]) : 200);     /* 時間ステップ数 */
  double coef  = 0.25;                                 /* (c*dt/dx)^2, 安定 */

  int n = L * L;
  double * up = (double *)calloc(n, sizeof(double));   /* u^{n-1} */
  double * cu = (double *)calloc(n, sizeof(double));   /* u^{n}   */
  double * nx = (double *)calloc(n, sizeof(double));   /* u^{n+1} */
  /* 初期条件: 中央にガウスの山, 初速 0 (= up と cu を同じにする)。境界は 0 のまま。 */
  double c0 = (L - 1) / 2.0, sig = L / 16.0;
  for (int i = 1; i < L - 1; i++)
    for (int j = 1; j < L - 1; j++) {
      double r2 = (i - c0) * (i - c0) + (j - c0) * (j - c0);
      double v = exp(-r2 / (2.0 * sig * sig));
      up[i * L + j] = v;
      cu[i * L + j] = v;
    }

  double t0 = omp_get_wtime();
  for (int t = 0; t < steps; t++) {
    /* 内部の各点を更新 (時間1ステップ進める) */
    // TODO: 内側の二重ループを #pragma omp parallel for collapse(2) で並列化せよ.
    for (int i = 1; i < L - 1; i++) {
      for (int j = 1; j < L - 1; j++) {
        double lap = cu[(i - 1) * L + j] + cu[(i + 1) * L + j]
                   + cu[i * L + (j - 1)] + cu[i * L + (j + 1)]
                   - 4.0 * cu[i * L + j];
        nx[i * L + j] = 2.0 * cu[i * L + j] - up[i * L + j] + coef * lap;
      }
    }
    /* up <- cu <- nx と時間を1つ進める (ポインタを回す) */
    double * tmp = up; up = cu; cu = nx; nx = tmp;
  }
  double elapsed = omp_get_wtime() - t0;

  /* 検算1: 最大振幅 (発散していなければ O(1) のまま)。
     検算2: 初期条件が i<->j 対称なので, 解も常に対称。max|u[i][j]-u[j][i]| は丸め誤差程度。 */
  double maxabs = 0.0, asym = 0.0;
  for (int i = 0; i < L; i++)
    for (int j = 0; j < L; j++) {
      double a = fabs(cu[i * L + j]);
      if (a > maxabs) maxabs = a;
      double s = fabs(cu[i * L + j] - cu[j * L + i]);
      if (s > asym) asym = s;
    }
  printf("L=%d, steps=%d: 最大振幅=%.4f, 対称性誤差=%.2e (≈0 なら正しい)\n",
         L, steps, maxabs, asym);
  printf("elapsed = %.3f sec\n", elapsed);
  free(up); free(cu); free(nx);
  return 0;
}

File: 15_pde_fdm/problems/01_wave/wave.f90
! 2D 波動方程式 u_tt = c^2 (u_xx + u_yy) を陽解法で時間発展させる。
! L×L の膜。四辺を固定 (u=0) し, 中央に山(ガウス)を置いて波が広がり反射する様子を計算する。
! 更新式 (5点ラプラシアン, coef=(c*dt/dx)^2, 安定条件 coef<=0.5):
!   u^{n+1} = 2 u^n - u^{n-1} + coef*(上+下+左+右 - 4*中央)
! 時間方向は前後ステップに依存するので逐次, 空間の二重ループを並列化する。
program wave
  use omp_lib
  implicit none
  integer :: L, steps, t, i, j
  real(8) :: coef, c0, sig, r2, v, lap, maxabs, asym, s, a, t0, elapsed
  real(8), allocatable, target :: arr1(:,:), arr2(:,:), arr3(:,:)
  real(8), pointer :: up(:,:), cu(:,:), nx(:,:), tmp(:,:)
  character(len=32) :: arg
  L = 257; steps = 200; coef = 0.25d0
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) L
  end if
  if (command_argument_count() >= 2) then
     call get_command_argument(2, arg); read (arg, *) steps
  end if

  allocate(arr1(0:L-1,0:L-1), arr2(0:L-1,0:L-1), arr3(0:L-1,0:L-1))
  arr1 = 0.0d0; arr2 = 0.0d0; arr3 = 0.0d0
  up => arr1; cu => arr2; nx => arr3
  ! 初期条件: 中央にガウスの山, 初速 0 (up と cu を同じに)。境界は 0 のまま。
  c0 = (L - 1) / 2.0d0; sig = L / 16.0d0
  do j = 1, L - 2
     do i = 1, L - 2
        r2 = (i - c0)**2 + (j - c0)**2
        v = exp(-r2 / (2.0d0 * sig * sig))
        up(i,j) = v
        cu(i,j) = v
     end do
  end do

  t0 = omp_get_wtime()
  do t = 1, steps
     ! 内部の各点を更新 (時間1ステップ進める)
     ! TODO: 内側の二重ループを !$omp parallel do collapse(2) private(lap) で並列化せよ.
     do j = 1, L - 2
        do i = 1, L - 2
           lap = cu(i-1,j) + cu(i+1,j) + cu(i,j-1) + cu(i,j+1) - 4.0d0 * cu(i,j)
           nx(i,j) = 2.0d0 * cu(i,j) - up(i,j) + coef * lap
        end do
     end do
     ! TODO: 上で始めた parallel do 領域を閉じる (!$omp end parallel do).
     ! up <- cu <- nx と時間を1つ進める (ポインタを回す)
     tmp => up; up => cu; cu => nx; nx => tmp
  end do
  elapsed = omp_get_wtime() - t0

  ! 検算1: 最大振幅。検算2: i<->j 対称性の誤差 (≈0 なら正しい)。
  maxabs = 0.0d0; asym = 0.0d0
  do j = 0, L - 1
     do i = 0, L - 1
        a = abs(cu(i,j))
        if (a > maxabs) maxabs = a
        s = abs(cu(i,j) - cu(j,i))
        if (s > asym) asym = s
     end do
  end do
  print "(a,i0,a,i0,a,f0.4,a,es9.2,a)", &
       "L=", L, ", steps=", steps, ": 最大振幅=", maxabs, ", 対称性誤差=", asym, " (≈0 なら正しい)"
  print "(a,f0.3,a)", "elapsed = ", elapsed, " sec"
end program wave
